### PR TITLE
win32: use native threading APIs instead of pthreads

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
* use fiber local storage if NO_PTHREADS is defined
* use critical sections instead of mutexes

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
Moderately contentious:

* fiber-local storage _is_ thread-local storage on all versions of Windows that matter
* the thread-local storage APIs (`Tls*`) are not available to store applications
* `NO_PTHREADS` is not the right thing to key off here
* We switched to critical sections because they're cheaper than mutexes. Mutexes can cross process boundaries and are a kernel affordance, and that incurs a round trip through kernel mode to manipulate them. CS are entirely in-process and usermode.

File: lock.h [hunk 1]
```diff
@@ -10,11 +10,11 @@
...
 #	include <windows.h>
 #undef BOOL
-typedef HANDLE mutex_t;
-#	define INIT_LOCK(x) x = CreateMutex(NULL, FALSE, NULL)
-#	define LOCK(x) WaitForSingleObject(*x, INFINITE)
-#	define UNLOCK(x) ReleaseMutex(*x)
-#	define DESTROY_LOCK(x) CloseHandle(*x)
+typedef CRITICAL_SECTION mutex_t;
+#	define INIT_LOCK(x) InitializeCriticalSection(&(x))
+#	define LOCK(x) EnterCriticalSection(x)
```
Comment (DHowett-MSFT) [inline lock.h]: critical sections can time out after about 10 minutes. this was deemed an acceptable cost.

File: arc.m [hunk 2]
```diff
@@ -14,9 +14,36 @@
...
+#define arc_tls_load pthread_getspecific
+#define TLS_CALLBACK(name) void name
+
+static inline pthread_key_t arc_tls_key_create(void(*cleanupFunction)(void*))
+{
+	pthread_key_t key;
+	pthread_key_create(&key, cleanupFunction);
+	return key;
+}
+#else // if defined(NO_PTHREADS)
```
Comment (davidchisnall) [inline arc.m]: This should be in some kind of `#ifdef` Windows thing - Windows isn't the only platform with no pthreads.  I suspect that for Windows this actually shouldn't be in the `#else` bit at all: even if we have Pthreads on Windows via the MinGW pthreads implementation, we'd probably prefer to use the native Windows APIs.

File: arc.m [hunk 3]
```diff
@@ -60,17 +87,13 @@ - (void)release;
 
 static inline struct arc_tls* getARCThreadData(void)
 {
-#ifdef NO_PTHREADS
-	return NULL;
-#else
-	struct arc_tls *tls = pthread_getspecific(ARCThreadKey);
+	struct arc_tls *tls = arc_tls_load(ARCThreadKey);
```
Comment (davidchisnall) [inline arc.m]: We still need to keep the `NO_PTHREADS` case for platforms that both are not Windows and don't support pthreads.

File: lock.h (hunk 1 above)
Comment (davidchisnall) [reply lock.h]: No objection to this, but I wondered if you'd done any profiling of Mutex vs CriticalSection vs SRW lock for workloads that you care about.

Comment (DHowett-MSFT) [reply lock.h]: The data there is, unfortunately, lost to the ages. @bviglietta made the switch for us ~2 years ago.

**Comment by DHowett-MSFT** [comment]
We now prefer Fls on Windows (even MinGW). Fls is available down to Vista/Server 2003.

**Comment by DHowett-MSFT** [comment]
Fixed an issue in `eh_personality.c` that came up building with NO_PTHREADS. It looks like ARC-compatible autorelease doesn't work at all with pthreads disabled (I'll file an issue on this).

Additionally, some Windows defines were preventing us building properly outside WinObjC; those are now mostly fixed.

**Pull request merged**